src/OpenLoco/src/Ui/WindowManager.cpp

```diff
@@ -1466,6 +1467,106 @@ namespace OpenLoco::Ui::WindowManager
         invalidateWidget(window->type, window->number, widgetIndex);
     }
 
+    // Allow mouse wheel scrolling to manipulate stepper widgets and tool sizes
+    static bool stepperWheelInput(Window& w, WidgetIndex_t widgetIndex, int32_t wheel)
+    {
+        auto widgetType = w.widgets[widgetIndex].type;
+
+        // If we aren't scrolling on a tool preview or stepper textbox, lower the widget index
+        // once or twice to try to find the parent widget. This allows scrolling on the
+        // increase/decrease buttons themselves, too.
+        int32_t attempts = 0;
+        while (widgetType != WidgetType::wt_3 && widgetType != WidgetType::textbox && widgetIndex > 0)
+        {
+            switch (widgetType)
+            {
+                case WidgetType::toolbarTab: // + and - for tool widget
+                case WidgetType::button:     // + and - for stepper widget
+                {
+                    if (attempts > 0)
+                    {
+                        // Verify that the previous button was of the same type
+                        auto previousType = w.widgets[widgetIndex + 1].type;
+                        if (previousType != widgetType)
+                        {
+                            return false;
+                        }
+                    }
+                    break;
+                }
+                default:
+                    // The widget type is not an increment or decrement button
+                    return false;
+            }
+
+            attempts++;
+            if (attempts > 2)
+            {
+                // We've looked two widgets up, and no tool preview or stepper widget was found
+                return false;
+            }
+
+            widgetIndex--;
+            widgetType = w.widgets[widgetIndex].type;
+        }
+
+        assert(w.widgets[widgetIndex + 1].type != WidgetType::end);
+        assert(w.widgets[widgetIndex + 2].type != WidgetType::end);
```

Comment:
If +1 is the end then +2 would be out of bounds here.

Reply:
True, but isn't the point of this assertion to make sure that doesn't happen? As in, if +1 is of WidgetType::end, the former assertion fails and so the latter assertion for +2 isn't even evaluated.

Reply:
I guess that's correct, however this only triggers for debug builds so we are not really guarding this case for release versions where this might read some arbitrary memory at the end not necessarily crashing, can't we bail if we reached the end?

Reply:
I think we've got that covered with lines 1519–1552 though. Would you agree?

Reply:
You might be right, this is quite complex to fully understand which is probably an indicator that this should be re-worked a bit.

Reply:
Yeah its definitely confusing code. Every time i look at it I come away less sure how it works. Afaik we have the following:

The input into this function is a widget index and it could be any widget index of the window.

A stepper is comprised of 3 widgets: wt_3/textbox, + toolbartab/button, - toolbartab/button. If widgetIndex is a +, - the first bit of code looks backwards until widgetIndex it is pointing at the wt_3/textbox (unless there has been some issue and its now 0). If wt_3/textbox not found after 2 steps then it returns.
If widgetIndex was 1 and wt_3/textbox was not found at 0 or 1 then we have an issue! And we continue with widgetIndex as 0 but without validating that +2 is allowed.

So at that point widgetIndex + 1 and + 2 should be safe (unless there was an issue where its 0). The second bit of code then confirms that widgetIndex + 1 and widgetIndex + 2 are indeed + toolbartab/button, - toolbartab/button.

So as long as the first bit of code is robust i think we are safe. (I know this isn't what i said before...)
And the asserts will never access invalid memory as the function will have returned under all the situations that might occur.